File: backend/app/schemas/compliance.py
"""
Compliance Schemas — Pydantic models for compliance session CRUD,
decision tree responses, and analysis results.
"""
from datetime import datetime
from typing import Optional, List
from uuid import UUID
from pydantic import BaseModel, Field


# ─── Session Schemas ───────────────────────────────────────────────────────

class ComplianceSessionCreate(BaseModel):
    """Create a new compliance analysis session"""
    client_name: str
    framework: str = "IFRS"
    agent_id: Optional[UUID] = None


class ComplianceSessionUpdate(BaseModel):
    """Update a compliance session"""
    client_name: Optional[str] = None
    framework: Optional[str] = None
    status: Optional[str] = None
    current_stage: Optional[int] = None
    selected_standards: Optional[List[str]] = None
    extracted_metadata: Optional[dict] = None
    analysis_results: Optional[dict] = None
    compliance_score: Optional[int] = None
    total_standards: Optional[int] = None
    total_questions: Optional[int] = None
    compliant_count: Optional[int] = None
    non_compliant_count: Optional[int] = None
    not_applicable_count: Optional[int] = None


class ComplianceSessionResponse(BaseModel):
    """Compliance session response"""
    id: UUID
    session_code: str
    client_name: str
    framework: str
    status: str
    current_stage: int
    financial_statements_filename: Optional[str] = None
    notes_filename: Optional[str] = None
    selected_standards: Optional[List[str]] = None
    total_standards: int = 0
    total_questions: int = 0
    extracted_metadata: Optional[dict] = None
    analysis_results: Optional[dict] = None
    compliance_score: Optional[int] = None
    compliant_count: int = 0
    non_compliant_count: int = 0
    not_applicable_count: int = 0
    messages: Optional[list] = None
    agent_id: Optional[UUID] = None
    created_by: UUID
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ComplianceSessionListResponse(BaseModel):
    """Lightweight session for list views"""
    id: UUID
    session_code: str
    client_name: str
    framework: str
    status: str
    current_stage: int
    compliance_score: Optional[int] = None
    created_at: datetime

    model_config = {"from_attributes": True}


# ─── Decision Tree Schemas ─────────────────────────────────────────────────

class DecisionTreeNode(BaseModel):
    """A node in the decision tree"""
    question: str
    guidance: Optional[str] = None
    yes_case: Optional[object] = None
    no_case: Optional[object] = None


class ComplianceItem(BaseModel):
    """A single compliance check item from a standard"""
    id: str
    section: str
    reference: str
    question: str
    question_type: Optional[str] = None
    source_question: Optional[str] = None
    source_trigger: Optional[str] = None
    context_required: Optional[str] = None
    original_question: Optional[str] = None
    decision_tree: Optional[dict] = None


class ComplianceStandard(BaseModel):
    """A compliance standard with its items"""
    section: str
    title: str
    description: Optional[str] = None
    item_count: int
    file_name: str


class ComplianceStandardDetail(BaseModel):
    """Full standard with all items"""
    section: str
    title: str
    description: Optional[str] = None
    items: List[ComplianceItem]


class StandardsSummary(BaseModel):
    """Summary of all available standards"""
    total_standards: int
    total_questions: int
    frameworks: List[str]
    standards: List[ComplianceStandard]


# ─── File Upload Response ──────────────────────────────────────────────────

class FileUploadResponse(BaseModel):
    """Response after file upload"""
    session_id: UUID
    financial_statements_uploaded: bool
    notes_uploaded: bool
    status: str
    message: str


# ─── Analysis Result Schemas ───────────────────────────────────────────────

class ComplianceResultItem(BaseModel):
    """A single compliance analysis result"""
    question_id: str
    standard: str
    section: str
    reference: str
    question: str
    status: str  # YES, NO, N/A, ERROR
    confidence: float = 0.0
    explanation: str = ""
    evidence: str = ""
    suggested_disclosure: str = ""
    decision_tree_path: List[str] = Field(default_factory=list)
    sequence: int = 1
    analysis_time_ms: int = 0
    error: Optional[str] = None


class AnalysisSummary(BaseModel):
    """Aggregated analysis results summary"""
    total: int
    compliant: int
    non_compliant: int
    not_applicable: int
    errors: int = 0
    compliance_score: int = 0
    avg_confidence: float = 0.0
    by_standard: Optional[dict] = None


class AnalysisResponse(BaseModel):
    """Full analysis response"""
    session_id: UUID
    status: str
    compliance_score: int
    summary: AnalysisSummary
    total_results: int
    analysis_time_seconds: float


class MetadataExtractionResponse(BaseModel):
    """Response from metadata extraction"""
    session_id: UUID
    metadata: dict
    status: str


class StandardSuggestionResponse(BaseModel):
    """Response from AI standard suggestion"""
    session_id: UUID
    suggested_standards: List[str]
    total_suggested: int


# ─── Chunk Management Schemas ──────────────────────────────────────────────

class ChunkPreviewItem(BaseModel):
    """A single chunk for preview display"""
    chunk_id: str
    chunk_index: int
    content: str
    taxonomy: str = "general"
    has_table: bool = False
    char_count: int = 0


class ChunkPreviewResponse(BaseModel):
    """Response with chunk previews for a session"""
    session_id: UUID
    total_chunks: int
    chunks: List[ChunkPreviewItem]
    taxonomy_summary: dict = Field(default_factory=dict)


class ChunkRevalidateRequest(BaseModel):
    """Request to re-validate/re-classify specific chunks"""
    chunk_ids: List[str] = Field(default_factory=list)
    reclassify: bool = False


class ChunkRevalidateResponse(BaseModel):
    """Response after chunk revalidation"""
    session_id: UUID
    revalidated_count: int
    taxonomy_changes: int
    status: str


# ─── Financial Statement Validation Schemas ────────────────────────────────

class FinancialValidationResult(BaseModel):
    """Result of financial statement validation"""
    is_valid: bool
    detected_statements: List[str] = Field(default_factory=list)
    missing_statements: List[str] = Field(default_factory=list)
    warnings: List[str] = Field(default_factory=list)
    confidence: float = 0.0


class FinancialValidationResponse(BaseModel):
    """Response from financial statement validation"""
    session_id: UUID
    validation: FinancialValidationResult
    status: str


# ─── Question Filter Schemas ──────────────────────────────────────────────

class QuestionFilterRequest(BaseModel):
    """Request to filter questions with optional NLP instructions"""
    standards: List[str]
    instructions: Optional[str] = None
    exclude_question_ids: List[str] = Field(default_factory=list)


class QuestionFilterResponse(BaseModel):
    """Response with filtered questions"""
    session_id: UUID
    total_before: int
    total_after: int
    filtered_questions: List[ComplianceItem]
    applied_instructions: Optional[str] = None


# ─── Chatbot Schemas ───────────────────────────────────────────────────────

class ChatConversationCreate(BaseModel):
    """Create a new chat conversation"""
    title: Optional[str] = None
    context_question_id: Optional[str] = None


class ChatConversationResponse(BaseModel):
    """Chat conversation response"""
    id: UUID
    session_id: UUID
    title: Optional[str] = None
    context_question_id: Optional[str] = None
    is_active: bool
    created_at: datetime
    updated_at: datetime
    message_count: int = 0

    model_config = {"from_attributes": True}


class ChatMessageCreate(BaseModel):
    """Send a new chat message"""
    content: str


class ChatMessageResponse(BaseModel):
    """Chat message response"""
    id: UUID
    conversation_id: UUID
    role: str
    content: str
    citations: Optional[List[dict]] = None
    created_at: datetime

    model_config = {"from_attributes": True}


class ChatSendResponse(BaseModel):
    """Response after sending a chat message (includes AI reply)"""
    user_message: ChatMessageResponse
    assistant_message: ChatMessageResponse


# ─── Saved Results Schemas ─────────────────────────────────────────────────

class SavedResultResponse(BaseModel):
    """Response for cached/saved analysis result"""
    id: UUID
    document_hash: str
    framework: str
    questions_hash: str
    results: dict
    metadata: Optional[dict] = None
    access_count: int
    last_accessed_at: datetime
    created_at: datetime

    model_config = {"from_attributes": True}


# ─── Job Tracking Schemas ─────────────────────────────────────────────────

class AnalysisProgressItem(BaseModel):
    """Single question progress"""
    question_id: str
    status: str
    error: Optional[str] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None


class AnalysisJobStatus(BaseModel):
    """Overall job status with per-question progress"""
    job_id: str
    session_id: UUID
    total_questions: int
    completed: int
    failed: int
    in_progress: int
    pending: int
    progress_percent: float
    items: List[AnalysisProgressItem] = Field(default_factory=list)


# ─── Health Check Schema ──────────────────────────────────────────────────

class HealthCheckResponse(BaseModel):
    """Health check response for compliance subsystem"""
    status: str
    azure_openai: str
    azure_search: str
    azure_doc_intelligence: str
    decision_trees_loaded: int
    total_questions: int

File: frontend/app/types/compliance.ts
/**
 * Compliance Module Types — TypeScript interfaces for the RAI Compliance Analysis Platform.
 *
 * Covers: sessions, analysis results, chunks, chatbot, standards, validation,
 * job tracking, and state machine stages.
 */

// ─── Enums / Constants ────────────────────────────────────────────────────

export type ComplianceSessionStatus =
  | "awaiting_upload"
  | "processing"
  | "metadata_review"
  | "framework_selection"
  | "standards_selection"
  | "context_preview"
  | "analyzing"
  | "completed"
  | "failed";

export type ComplianceFramework = "IFRS" | "US GAAP" | "Ind AS";

export type ComplianceResultStatus =
  | "YES"
  | "NO"
  | "N/A"
  | "PENDING"
  | "ERROR";

export type ChatMessageRole = "user" | "assistant" | "system";

export type AnalysisProgressStatus =
  | "pending"
  | "in_progress"
  | "completed"
  | "failed";

/**
 * Ordered stage definitions for the compliance analysis state machine.
 * Each stage has a numeric key, label, description, and associated status.
 */
export const COMPLIANCE_STAGES = {
  1: {
    label: "Upload Documents",
    description: "Upload financial statements and notes",
    status: "awaiting_upload" as ComplianceSessionStatus,
  },
  2: {
    label: "Metadata Review",
    description: "Review extracted document metadata",
    status: "metadata_review" as ComplianceSessionStatus,
  },
  3: {
    label: "Framework Selection",
    description: "Select compliance framework (IFRS, US GAAP, Ind AS)",
    status: "framework_selection" as ComplianceSessionStatus,
  },
  4: {
    label: "Standards Selection",
    description: "Choose which standards to analyze",
    status: "standards_selection" as ComplianceSessionStatus,
  },
  5: {
    label: "Context Preview",
    description: "Preview document chunks and validate coverage",
    status: "context_preview" as ComplianceSessionStatus,
  },
  6: {
    label: "Analysis",
    description: "AI-powered compliance analysis in progress",
    status: "analyzing" as ComplianceSessionStatus,
  },
  7: {
    label: "Results",
    description: "Review compliance analysis results",
    status: "completed" as ComplianceSessionStatus,
  },
} as const;

export type StageNumber = keyof typeof COMPLIANCE_STAGES;

// ─── Session Types ────────────────────────────────────────────────────────

export interface ComplianceSession {
  id: string;
  session_code: string;
  client_name: string;
  framework: string;
  status: ComplianceSessionStatus;
  current_stage: number;
  financial_statements_filename: string | null;
  notes_filename: string | null;
  selected_standards: string[] | null;
  total_standards: number;
  total_questions: number;
  extracted_metadata: ExtractedMetadata | null;
  analysis_results: AnalysisResults | null;
  compliance_score: number | null;
  compliant_count: number;
  non_compliant_count: number;
  not_applicable_count: number;
  messages: ChatLogMessage[] | null;
  agent_id: string | null;
  created_by: string;
  created_at: string;
  updated_at: string;
}

export interface ComplianceSessionListItem {
  id: string;
  session_code: string;
  client_name: string;
  framework: string;
  status: ComplianceSessionStatus;
  current_stage: number;
  compliance_score: number | null;
  created_at: string;
}

export interface ComplianceSessionCreate {
  client_name: string;
  framework?: string;
  agent_id?: string;
}

export interface ComplianceSessionUpdate {
  client_name?: string;
  framework?: string;
  status?: string;
  current_stage?: number;
  selected_standards?: string[];
  extracted_metadata?: Record<string, unknown>;
  analysis_results?: Record<string, unknown>;
  compliance_score?: number;
  total_standards?: number;
  total_questions?: number;
  compliant_count?: number;
  non_compliant_count?: number;
  not_applicable_count?: number;
}

// ─── Metadata Types ───────────────────────────────────────────────────────

export interface ExtractedMetadata {
  company_name?: string;
  reporting_period?: string;
  reporting_year?: string;
  currency?: string;
  industry?: string;
  auditor?: string;
  reporting_framework?: string;
  consolidated?: boolean;
  interim?: boolean;
  document_type?: string;
  key_accounting_policies?: string[];
  error?: string;
}

// ─── Analysis Result Types ────────────────────────────────────────────────

export interface ComplianceResultItem {
  question_id: string;
  standard: string;
  section: string;
  reference: string;
  question: string;
  status: ComplianceResultStatus;
  confidence: number;
  explanation: string;
  evidence: string;
  suggested_disclosure: string;
  decision_tree_path: string[];
  context_used: string[];
  sequence: number;
  analysis_time_ms: number;
  error: string | null;
  // Override fields (present when overridden)
  original_status?: string;
  override_reason?: string;
  overridden_by?: string;
}

export interface AnalysisSummary {
  total: number;
  compliant: number;
  non_compliant: number;
  not_applicable: number;
  errors: number;
  compliance_score: number;
  avg_confidence: number;
  by_standard?: Record<
    string,
    {
      total: number;
      compliant: number;
      non_compliant: number;
      not_applicable: number;
      errors: number;
      avg_confidence: number;
    }
  >;
}

export interface AnalysisResults {
  summary: AnalysisSummary;
  results: ComplianceResultItem[];
  document_hash: string;
  analysis_time_seconds: number;
}

export interface AnalysisResponse {
  session_id: string;
  status: string;
  compliance_score: number;
  summary: AnalysisSummary;
  total_results: number;
  analysis_time_seconds: number;
}

// ─── Streaming Types ──────────────────────────────────────────────────────

export interface StreamEvent {
  type: "status" | "progress" | "result" | "complete" | "error";
  data: Record<string, unknown>;
}

export interface StreamProgress {
  total_questions: number;
  completed_questions: number;
  percentage: number;
  current_standard: string;
  current_question: string;
  phase: string;
  errors: string[];
}

// ─── Decision Tree / Standards Types ──────────────────────────────────────

export interface ComplianceItem {
  id: string;
  section: string;
  reference: string;
  question: string;
  question_type?: string;
  source_question?: string;
  source_trigger?: string;
  context_required?: string;
  original_question?: string;
  decision_tree?: Record<string, unknown>;
}

export interface ComplianceStandard {
  section: string;
  title: string;
  description?: string;
  item_count: number;
  file_name: string;
}

export interface StandardsSummary {
  total_standards: number;
  total_questions: number;
  frameworks: string[];
  standards: ComplianceStandard[];
}

export interface StandardDetail {
  section: string;
  title: string;
  description?: string;
  items: ComplianceItem[];
}

// ─── Chunk Types ──────────────────────────────────────────────────────────

export interface ChunkPreviewItem {
  chunk_id: string;
  chunk_index: number;
  content: string;
  taxonomy: string;
  has_table: boolean;
  char_count: number;
}

export interface ChunkPreviewResponse {
  session_id: string;
  total_chunks: number;
  chunks: ChunkPreviewItem[];
  taxonomy_summary: Record<string, number>;
}

export interface ChunkRevalidateRequest {
  chunk_ids?: string[];
  reclassify?: boolean;
}

export interface ChunkRevalidateResponse {
  session_id: string;
  revalidated_count: number;
  taxonomy_changes: number;
  status: string;
}

// ─── Financial Validation Types ───────────────────────────────────────────

export interface FinancialValidationResult {
  is_valid: boolean;
  detected_statements: string[];
  missing_statements: string[];
  warnings: string[];
  confidence: number;
}

export interface FinancialValidationResponse {
  session_id: string;
  validation: FinancialValidationResult;
  status: string;
}

// ─── Question Filter Types ───────────────────────────────────────────────

export interface QuestionFilterRequest {
  standards: string[];
  instructions?: string;
  exclude_question_ids?: string[];
}

export interface QuestionFilterResponse {
  session_id: string;
  total_before: number;
  total_after: number;
  filtered_questions: ComplianceItem[];
  applied_instructions?: string;
}

// ─── Chatbot Types ────────────────────────────────────────────────────────

export interface ChatConversation {
  id: string;
  session_id: string;
  title: string | null;
  context_question_id: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  message_count: number;
}

export interface ChatMessage {
  id: string;
  conversation_id: string;
  role: ChatMessageRole;
  content: string;
  citations: ChatCitation[] | null;
  created_at: string;
}

export interface ChatCitation {
  text: string;
  source: string;
}

export interface ChatSendResponse {
  user_message: ChatMessage;
  assistant_message: ChatMessage;
}

// ─── Saved Results / Cache Types ──────────────────────────────────────────

export interface SavedResult {
  id: string;
  document_hash: string;
  framework: string;
  questions_hash: string;
  results: Record<string, unknown>;
  metadata: Record<string, unknown> | null;
  access_count: number;
  last_accessed_at: string;
  created_at: string;
}

// ─── Job Tracking Types ──────────────────────────────────────────────────

export interface AnalysisProgressItem {
  question_id: string;
  status: AnalysisProgressStatus;
  error: string | null;
  started_at: string | null;
  completed_at: string | null;
}

export interface AnalysisJobStatus {
  job_id: string;
  session_id: string;
  total_questions: number;
  completed: number;
  failed: number;
  in_progress: number;
  pending: number;
  progress_percent: number;
  items: AnalysisProgressItem[];
}

// ─── Health Check Types ──────────────────────────────────────────────────

export interface ComplianceHealthCheck {
  status: string;
  azure_openai: string;
  azure_search: string;
  azure_doc_intelligence: string;
  decision_trees_loaded: number;
  total_questions: number;
}

// ─── Chat Log (session-level messages) ───────────────────────────────────

export interface ChatLogMessage {
  role: string;
  content: string;
  timestamp: string;
}

// ─── File Upload Types ───────────────────────────────────────────────────

export interface FileUploadResponse {
  session_id: string;
  financial_statements_uploaded: boolean;
  notes_uploaded: boolean;
  status: string;
  message: string;
}

// ─── Metadata Extraction Response ────────────────────────────────────────

export interface MetadataExtractionResponse {
  session_id: string;
  metadata: ExtractedMetadata;
  status: string;
}

// ─── Standard Suggestion Response ────────────────────────────────────────

export interface StandardSuggestionResponse {
  session_id: string;
  suggested_standards: string[];
  total_suggested: number;
}
